ktor-server/ktor-server-config-yaml/nix/src/io/ktor/server/config/yaml/YamlConfigNix.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.server.config.yaml
+
+import io.ktor.server.config.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.pool.*
+import kotlinx.cinterop.*
+import net.mamoe.yamlkt.*
+import platform.posix.*
+
+private fun init() {
+    addConfigLoader(YamlConfigLoader())
+}
+
+@Suppress("DEPRECATION")
+@OptIn(ExperimentalStdlibApi::class)
+@EagerInitialization
+private val initHook = init()
+
+public actual fun YamlConfig(path: String?): YamlConfig? {
+    val resolvedPath = when {
+        path != null && path.endsWith(".yaml") -> path
+        path == null && access("application.yaml", F_OK) == 0 -> "application.yaml"
+        else -> null
+    } ?: return null
+    val content = readFile(resolvedPath)
+    val yaml = Yaml.decodeYamlFromString(content) as? YamlMap
+        ?: throw ApplicationConfigurationException("$resolvedPath should be a YAML dictionary")
+    return YamlConfig(yaml).apply { checkEnvironmentVariables() }
+}
+
+@OptIn(UnsafeNumber::class)
+private fun readFile(path: String): String {
+    val fd = fopen(path, "rb") ?: throw ApplicationConfigurationException("Can not read $path")
+    val bytes = ByteArrayPool.borrow()
+    val size = bytes.size
+    var read: Int
+    val packet = buildPacket {
+        do {
+            read = bytes.usePinned { pinned ->
+                fread(pinned.addressOf(0), 1.convert(), size.convert(), fd)
```

Comment:
`ferror` and  `feof` should be used to check the result

Reply:
I would extract `fread` to a separate method like `readFileChunk(descriptor, array, startIndex, endIndex)`